src/registry.rs

```diff
@@ -0,0 +1,208 @@
+    IncorrectInterface,
+}
+
+/// State object associated with the registry handling for smithay's client toolkit.
+#[derive(Debug)]
+pub struct RegistryHandle {
+    registry: wl_registry::WlRegistry,
+    cached_globals: HashMap<u32, CachedGlobal>,
+}
+
+impl RegistryHandle {
+    /// Creates a new registry handle.
+    pub fn new(registry: wl_registry::WlRegistry) -> RegistryHandle {
+        RegistryHandle { registry, cached_globals: HashMap::new() }
+    }
+
+    /// Binds a global, returning a new object associated with the global.
+    ///
+    /// This function may be used for any global, but should be avoided if the global being bound may be used
+    /// by multiple modules of smithay's client toolkit. If multiple modules need a global, use
+    /// [`RegistryHandle::bind_cached`] instead.
+    pub fn bind_once<I, D, U>(
+        &mut self,
+        cx: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        name: u32,
+        version: u32,
+        udata: U,
+    ) -> Result<I, BindError>
+    where
+        D: Dispatch<I, UserData = U> + 'static,
+        I: Proxy + 'static,
+        U: Send + Sync + 'static,
+    {
+        if let Entry::Occupied(entry) = self.cached_globals.entry(name) {
+            let cached = entry.get();
+
+            log::warn!(
+                target: "sctk",
+                "RegistryHandle::bind_once used to bind cached global {} (name: {})",
+                cached.interface,
+                cached.name
+            );
+        }
+
+        let global = self.registry.bind::<I, _>(cx, name, version, qh, udata)?;
+
+        log::debug!(target: "sctk", "Bound new global [{}] {} v{}", name, I::interface().name, version);
+
+        Ok(global)
+    }
+
+    /// Binds a global, caching the bound global for other modules of smithay's client toolkit to use.
+    ///
+    /// This function is primarily intended for globals which multiple modules may need to access, such as a
+    /// `wl_output`.
+    ///
+    /// The closure passed into the function will be invoked to obtain the version of the global to bind and
+    /// the user data associated with the global if the global has not been bound yet.
+    pub fn bind_cached<I, D, F, U>(
+        &mut self,
+        cx: &mut ConnectionHandle,
+        qh: &QueueHandle<D>,
+        name: u32,
+        f: F,
+    ) -> Result<I, BindError>
+    where
+        D: Dispatch<I, UserData = U> + 'static,
+        I: Proxy + 'static,
+        U: Send + Sync + 'static,
+        F: FnOnce() -> (u32, U),
+    {
+        match self.cached_globals.get(&name) {
+            Some(cached) => {
+                // Ensure the requested interface is the same.
+                if I::interface().name == cached.interface {
+                    // Create a new handle for the existing global.
+                    Ok(I::from_id(cx, cached.id.clone())?)
+                } else {
+                    Err(BindError::IncorrectInterface)
+                }
+            }
```

Comment:
We need to handle the case where the underlying object is no longer alive (for example it was destroyed by something else). This is when `I::from_id` would return an error. In that case we should instead `log::warn!()` it and bind the global anew.